### PR TITLE
fix: Use getDataProvider().getId(item) instead of item.hashCode()

### LINKED ISSUE #1270: TreeGrid breaks when items have the same hashCode
Sample code:
```
@Route(value = "")
public class TreeTest extends VerticalLayout {
    private final TreeGrid<String> treeGrid = new TreeGrid<>();

    public TreeTest() {
        TreeData<String> stringTreeData = new TreeData<>();
        // "Aa".hashCode() == "BB".hashCode()
        stringTreeData.addItem(null, "Aa");
        stringTreeData.addItem("Aa", "BB");
        treeGrid.setDataProvider(new TreeDataProvider<>(stringTreeData));
        treeGrid.addHierarchyColumn(s -> s);
        treeGrid.expand("Aa");
        add(treeGrid);
    }
}
```

This fails on the client:
Selection_358

If one of the strings is changed, everything works fine.

In our application we have this problem with a TreeGrid that displays JPA entites. The user can create new entries in the tree (that do not have an identifier assigned by jpa yet), so they have the same hash code. We want to save those new entities only when the user clicks 'Save'.

It would also break with entities that follow the recommendation of the hibernate developer advocate for implementing hash code: https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier/

Tested with Vaadin 13.0.3.

### PR DESCRIPTION
**Pull request by TatuLund**
Fixes #1270 
Fixes https://github.com/vaadin/flow-components/issues/2383
Fixes https://github.com/vaadin/flow-components/issues/464

### CODE REVIEW

**Comment by tomivirkki** [comment]
> fixes: #2411

points to this PR

**Comment by TatuLund** [comment]
> points to this PR

@tomivirkki thanks, fixed

File: vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java [hunk 1]
```diff
@@ -283,7 +283,6 @@ public void expanded_nodes_stay_expanded_when_parent_expand_state_is_toggled() {
         Assert.assertEquals(9, getTreeGrid().getRowCount());
 
         getTreeGrid().expandWithClick(7);
-        getTreeGrid().expandWithClick(8);
```
Comment (tomivirkki) [inline vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java:286]: Does this mean that an existing test breaks with the change in place?

There should also be a new test that 1. fails without the change 2. passes with the change

Comment (TatuLund) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java:286]: My understanding is that existing test is faulty and this line should not be there. The bug I am fixing has allowed the existing test to pass. I have asked internally confirmation about this from the original author of the test.

Comment (TatuLund) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java:286]: > There should also be a new test that 1. fails without the change 2. passes with the change

Done

Comment (tomivirkki) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java:286]: After the change, the Grid on the test page behaves quite unexpectedly in general. There may be applications out there that also rely on the current item identification logic (based on hashCode) and after we'd release this change, they could also start acting weird.

After the change:

https://user-images.githubusercontent.com/1222264/150156899-c0737864-1a7f-4a13-b793-2fa076e57026.mp4

Before the change:

https://user-images.githubusercontent.com/1222264/150156919-a3b2cbb4-60c8-4c69-a274-3f682980f055.mp4

Also, if we change the item ID logic in `TreeGrid`, shouldn't we also change it in `Grid` (and any other component that uses items) for consistency?

Comment (TatuLund) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java]: Thanks for this, it gave me an idea to check one thing.

Comment (TatuLund) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java]: I think the problem is in test page

https://github.com/vaadin/flow-components/blob/fix2383/vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/main/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesPage.java#L170

There is a bit rigged identity logic in the test entity, and there are more than one bean with id == null. And that is actually causing the problem. Thus the test bean is not realistic and this should be fixed.

Comment (tomivirkki) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java:286]: Are you still working on this? The fix to `TreeGrid` lgtm but this change to the test case isn't correct. As you suggested, the test fix should be done to the `HierarchicalTestBean` identity logic.

Comment (TatuLund) [reply vaadin-grid-flow-parent/vaadin-grid-flow-integration-tests/src/test/java/com/vaadin/flow/component/treegrid/it/TreeGridBasicFeaturesIT.java]: Lets see, if it works now

**Review by tomivirkki — APPROVED** [review]

**Pull request merged**